solutions/usaco-596.mdx

```diff
@@ -97,3 +97,76 @@ void ff(int i, int j)
 </CPPSection>
 
 </LanguageSection>
+	
+## Alternate Faster Solution
+
+It is possible to make the following 2 observations:
+
+1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate. For example, if there are 2 regions, there must be 2 gates.
+2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous closed region if all of these conditions are met:
+	* FJ walks along an edge from point $a$ to $b$
+	* He did not walk from $a$ to $b$ in the past
+	* He did not walk from $b$ to $a$ in the past
+	* $b$ is a point he had already visited
+
+Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is about 10 times faster than flood filling due to $\log N$ lookup and insertion.
+	
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	freopen("gates.in", "r", stdin);
+	freopen("gates.out", "w", stdout);
+	
+	ios::sync_with_stdio(0);
+	cin.tie(0);
+
+	int n; cin >> n;
+	string s; cin >> s;
+	
+	set<pair<pair<int, int>, pair<int, int>>> visedge;
+	set<pair<int, int>> visnode;
+	int ans = 0;
+	pair<int, int> prev = {0, 0};
```

Comment:
```suggestion
	pair<int, int> prev{0, 0};
```
more concise

Reply:
The equals is more clear

Reply:
not sure how it's more clear since they mean the exact same thing??

Reply:
not really? think list initialization is fairly standard?